### DISCUSSION TITLE
Completion date: Optional or not?

### DISCUSSION THREAD

**Discussion by ptosi** [todotxt/todo.txt#52] [Q&A]
The [illustration of the `todo.txt` format](../blob/b5c5f5ca/description.png) states

> Optional - Completion Date

while the [definition of the format for complete tasks](../blob/b5c5f5ca/todo.txt) states that the completion date is required:

> With the **completed date (required)**, if you've used the prepended date (optional), you can calculate how many days it took to complete a task.

Therefore, is the "Optional" supposed to mean "depending on the completion status of the task" or is this a discrepancy? In particular, can a complete task omit its completion date?

**Comment by davidstosik** [comment]
I'm confused as well, and here are two more lines that add to the confusion (emphasis mine):

> Optional -- Creation date (**must be specified if completion date is**)

> **If** you’ve prepended the creation date to your task, on completion it will appear directly after the completion date.

So far we have:
- creation date is optional
- completion date must be provided when the task is completed (`x`)
- creation date must be provided when the completion date is

## Problem 💥 

Now, what happens when I complete the task `call mom` (nothing more on the todo.txt line, especially no creation date, as it is optional)?
1. Completed date is required, so I need to add it:
    ```
    x 2019-07-03 call mom
    ```
2. But now, as completion date is provided, creation date must be specified too, but I don't have it!
    ```
    x 2019-07-03 ????-??-?? call mom
    ```

## Interpretation 💭 

My take, which does not perfectly match the spec (no logical interpretation can, in my opinion) is that:
- creation date is optional (_never_ required): if it is optional on incomplete tasks, it cannot reasonably be required on complete tasks
- completion date does not make sense for incomplete tasks
- completion date is required on completed tasks
- creation date may stick to completed tasks, or be removed, to the writer's liking

### Valid ✅ 
- `call mom`: when incomplete, no date is required
- `2019-07-01 call mom`: the provided date is the creation date
- `x 2019-07-03 call mom`: the provided date is the completion date
- `x 2019-07-03 2019-07-01 call mom`: the provided dates are, in order, completion then creation

### Invalid ❌ 
- `2019-07-03 2019-07-01 call mom`: the task is incomplete, so can't have a completion date
- `x call mom`: a completed task needs _at least_ a completion date

### Arguable point

My personal interpretation states that completion date is required when a task is complete. Some people might be happy to just `x` out a completed task without having to input more text.

**Comment by nicojx** [comment]
My take is that the completion date is (and should be) optional. For two reasons:
* `todo.txt` is aiming to be minimalist and flexible. The use of dates can be useful, but it is not for everyone, and if you use a simple text editor and don't _need_ dates, you should be able to do without. What's wrong with `x some task` if you just want something quick and easy?
* That's what it says: 

> If a task starts with an `x` (case-sensitive and lowercase) followed directly by a space, it is marked as complete.`

Yes, there's also
> With the completed date (required), if you've used the prepended date (optional), you can calculate how many days it took to complete a task.

But that could (and I think should) be interpreted as being required _if you use dates_, particularly if you use creation dates.

It would actually be helpful to clarify this, as different apps interpret this differently (most use the minimalist approach, but some don't), and it really isn't very clear in the `primer`.
